osutil/disks/gpt.go

```diff
@@ -0,0 +1,134 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package disks
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"os"
+)
+
+type GPTLBA uint64
+type GPTGUID [16]byte
+
+type GPTHeader struct {
+	Signature      [8]byte
+	Revision       uint32
+	HeaderSize     uint32
+	HeaderCRC      uint32
+	Reserved       uint32
+	CurrentLBA     GPTLBA
+	AlternateLBA   GPTLBA
+	FirstUsableLBA GPTLBA
+	LastUsableLBA  GPTLBA
+	DiskGUID       [16]byte
+	EntriesLBA     GPTLBA
+	NEntries       uint32
+	EntrySize      uint32
+	EntriesCRC     uint32
+}
+
+func LoadGPTHeader(devfd *os.File) (GPTHeader, error) {
+	var header GPTHeader
+	rawHeader := make([]byte, 512)
+	read, err := devfd.Read(rawHeader)
+	if err != nil {
+		return header, err
+	}
+
+	if read < binary.Size(header) {
+		return header, fmt.Errorf("Not enough data was read")
+	}
+
+	buf := bytes.NewReader(rawHeader[:read])
+	binary.Read(buf, binary.LittleEndian, &header)
+
+	if !bytes.Equal(header.Signature[:], []byte("EFI PART")) {
+		return header, fmt.Errorf("GPT Header does not start with magic string")
+	}
+
+	if header.Revision != 1 << 16 {
+		return header, fmt.Errorf("GPT header revision is not 1.0")
+	}
+
+	if int(header.HeaderSize) < binary.Size(header) {
+		return header, fmt.Errorf("GPT header size is too small")
+	}
+
+	if int(header.HeaderSize) > read {
+		return header, fmt.Errorf("GPT header size is too big")
+	}
+
+	for i := 0; i < 4; i++ {
+		rawHeader[8+4+4+i] = 0
+	}
+	crc := crc32.ChecksumIEEE(rawHeader[:header.HeaderSize])
+	if crc != header.HeaderCRC {
+		return header, fmt.Errorf("GPT header CRC32 checksum failed: %v != %v", crc, header.HeaderCRC)
+	}
+
+	return header, nil
+}
+
+func ReadGPTHeader(device string) (GPTHeader, error) {
+	var header GPTHeader
+	devfd, err := os.Open(device)
+	if err != nil {
+		return header, err
+	}
+	defer devfd.Close()
+
+	_, err = devfd.Seek(512, 0)
+	if err != nil {
+		return header, err
+	}
+
+	header, main_err := LoadGPTHeader(devfd)
+	if main_err != nil {
+		// Read the backup header
+		_, err := devfd.Seek(512, 2)
+		if err != nil {
+			return header, main_err
+		}
+
+		header, err := LoadGPTHeader(devfd)
+		if err != nil {
+			return header, main_err
+		}
+	}
+
+	return header, nil
+}
+
+func CalculateLastUsableLBA(device string) (uint64, error) {
+	header, err := ReadGPTHeader(device)
+	if err != nil {
+		return 0, err
+	}
+	sectors, err := blockDeviceSizeInSectors(device)
+	if err != nil {
+		return 0, err
+	}
+	entriessize := uint64(header.NEntries) * uint64(header.EntrySize)
+	entriessectors := (entriessize + 511) / 512
+	return sectors - entriessectors - 2, nil
+}
```

Comment:
could you add some comments explaining these operations? AFAICT you're just counting how many headers are used by the gpt headers / backup headers, and then subtracting that from the total number of sectors, but it's a bit unclear to my why you are doing `(entriessize + 511) / 512` for example

Reply:
Done.